### PR TITLE
Introducing new options for ties, and a general overhaul of tie layout

### PR DESCRIPTION
**Pull request by mike-spa**
Resolves: #18685 

This PR implements the feature requested in #18685. In order to do that, a deep overhaul of the tie layout code became necessary, as the existing code was too complex and inadequate to achieve this.

Here's a (probably incomplete) summary of the changes.
- Endpoints position is (slightly) changed and, if available, approximates the optical center of noteheads using smufl cutouts.
- Default curvature of ties (especially short and short-medium) is increased.
- The logic for vertical adjustment, which checks for bad endpoint and arc intersections with staff lines, is improved and avoids a few undesirable situations we had before.
- New style setting for inside/outside style (which defaults to "outside" as requested), as well as individual property override.
- Whenever possible (and if short enough) ties will fit within a space without intersecting staff lines.
- Very short ties are thinned.
- The annoying glitch when drag-editing ties has been fixed.
- New logic to deal with ledger lines.

### CODE REVIEW

File: src/engraving/libmscore/property.cpp
Comment (Jojo-Schmitz) [inline src/engraving/libmscore/property.cpp:408]: Nitpick: better reduce the space bevore that `DUMMY...` to make it align with the others

File: src/engraving/style/styledef.cpp
Comment (Jojo-Schmitz) [inline src/engraving/style/styledef.cpp:483]: Nore nitpicks: here too a better alignment seems due

File: src/engraving/rendering/dev/chordlayout.cpp
Comment (cbjeukendrup) [inline src/engraving/rendering/dev/chordlayout.cpp:674]: Now the `system` and `stick` parameters are unused and can be removed. (And that will require giving the two overloads of `ChordLayout::layoutSpanners` different names, which seems a good idea anyway.)

Actually, why did you decide to move this code out of this method?

**Review by cbjeukendrup — COMMENTED** [review]
Here's a first round of comments, mainly about the 'supporting code'. Now I'm going to look into the new layout code.

File: src/engraving/dom/tie.h
Comment (cbjeukendrup) [inline src/engraving/dom/tie.h:41]: This does not have the desired effect. You're probably thinking of `std::vector`'s constructor, but `std::array` does not have a similar thing. It only has "aggregate initialisation", meaning that you'll have essentially the same effect as with plain old C arrays:
```c++
    … m_adjustOffsets = { 
        PointF(),
        <default_initialised_by_compiler>,
        <default_initialised_by_compiler>,
        ⋮
        <default_initialised_by_compiler> // static_cast<size_t>(Grip::GRIPS) - 1 times
    };
```
Luckily, for structs and classes, `<default_initialised_by_compiler>` will be `PointF()`, so it will work correctly. But that means you could also not have specified anything at all, so that the first entry is also `<default_initialised_by_compiler>`; I'd argue that it's better to do that.
```suggestion
    std::array<PointF, static_cast<size_t>(Grip::GRIPS)> m_adjustOffsets;
```

File: src/engraving/dom/tie.cpp
Comment (cbjeukendrup) [inline src/engraving/dom/tie.cpp:1080]: Is this really the desired formula, or should it be
```c++
    return startN && endN && startN->chord()->vStaffIdx() != endN->chord()->vStaffIdx();
```
or, like Slur::isCrossStaff
```c++
    return startCR() && endCR()
           && (startCR()->staffMove() != 0 || endCR()->staffMove() != 0
               || startCR()->vStaffIdx() != endCR()->vStaffIdx());
```

**Comment by cbjeukendrup** [comment]
Let's not worry about those unit tests; I'll just review #18722 asap so that we can merge that first and then we simply rebase this one. (I guess that was your plan already :) )

File: src/engraving/dom/tie.cpp
Comment (mike-spa) [reply src/engraving/dom/tie.cpp:312]: That's quite a time consuming refactoring step, and I don't think it's worth doing it now for a method that is going to be deleted anyway. There were a few things (and there still is, to less extent) in Tie layout that were not moved to the dedicated layout class. Once we replace dev layout into stable, I've made a note to myself to go back and delete these methods that aren't used anymore.

File: src/engraving/rendering/dev/chordlayout.cpp
Comment (mike-spa) [reply src/engraving/rendering/dev/chordlayout.cpp:674]: Cause I meant to delete it (I did now)

File: src/engraving/dom/tie.cpp
Comment (mike-spa) [reply src/engraving/dom/tie.cpp:1080]: Yes, it's correct as it is (we consider it "cross" even if both endpoints are moved to the same staff). Actually, I think the last OR condition in Slur::isCrossStaff is redundant (it can't be true if at least one of the first two isn't) but I'll verify that another time

File: src/engraving/rendering/dev/slurtielayout.cpp
Comment (mike-spa) [reply src/engraving/rendering/dev/slurtielayout.cpp:1001]: This isn't code I wrote, I just moved it out of the way. Actually, I don't even know why it exists, I don't think I've found any situation when this is called, so I wouldn't know what that if-else is supposed to do. I'll just correct the obvious error and leave it as is

**Comment by mike-spa** [comment]
@cbjeukendrup besides my comment replies, all done, thanks!

**Review by cbjeukendrup — APPROVED** [review]
I found a couple more things, but decided to fix them my self in order not to bother you again :)

**Comment by cbjeukendrup** [comment]
Hm, it looks like the tests still need one more update. @mike-spa could you take care of that? (you don't need to wait for a new approval after that, since it only seems to involve test reference files)

**Pull request merged**
